### PR TITLE
spirv-opt: Add a folding rule for `OpBitReverse`.

### PR DESCRIPTION
**Pull request by luciechoi**
Add a folding rule for `OpBitReverse` on scalar and vector types.

Noticed we are missing this optimization while working on https://github.com/microsoft/DirectXShaderCompiler/issues/7680

**Additionally, fix various constant folding errors for `OpBitCast` to handle null constants as well as lower-bit to higher-bit  integer conversions.** Previously, it was throwing error for e.g. `OpBitcast %int %v2ushort_1_null`.

### CODE REVIEW

File: test/opt/fold_test.cpp [hunk 1]
```diff
@@ -1471,7 +1483,16 @@ ::testing::Values(
...
+        2, {2,65531}),
+    // Test case 7: fold bitreverse of a vector of uint
+    InstructionFoldingCase<std::vector<uint32_t>>(
+      Header() + "%main = OpFunction %void None %void_func\n" +
+          "%main_lab = OpLabel\n" +
+          "%n = OpVariable %_ptr_int Function\n" +
+          "%load = OpLoad %int %n\n" +
+          "%2 = OpBitReverse %v2uint %v2uint_1_0x0000ffff\n" +
+          "OpReturn\n" + "OpFunctionEnd",
+      2, {2147483648, 0xffff0000})
```
Comment (s-perron) [inline test/opt/fold_test.cpp]: Use hex for 2147483648. It is easier to verify the bits are in the right place when reading in hex.

Comment (luciechoi) [reply test/opt/fold_test.cpp]: Done.

**Comment by luciechoi** [comment]
> The implementation looks good, but it could use more tests. Just to have a little more coverage in case someone decides to change it.

Thank you, I added more test cases and modified the folding logic to handle null constants ~(or PLMK if it's better to fold them as well). We can modify the [util](https://github.com/KhronosGroup/SPIRV-Tools/blob/745e753fd68bc59d170bb812ba3b0b1123f66fe0/source/opt/folding_rules.cpp#L185), but this is also used for folding [BitCast](https://github.com/KhronosGroup/SPIRV-Tools/blob/745e753fd68bc59d170bb812ba3b0b1123f66fe0/source/opt/folding_rules.cpp#L2226C9-L2226C50), where I'm not sure about the exact behavior for `OpConstantNull`.~

File: source/opt/folding_rules.cpp [hunk 2]
```diff
@@ -2259,6 +2259,18 @@ FoldingRule BitReverseScalarOrVector() {
         GetWordsFromNumericScalarOrVectorConstant(const_mgr, constants[0]);
     if (words.size() == 0) return false;
 
+    if (constants[0]->AsNullConstant() ||
```
Comment (s-perron) [inline source/opt/folding_rules.cpp]: Null is just `0`. If the value is a null constant, then you can just return the null constant.

File: source/opt/folding_rules.cpp [hunk 3]
```diff
@@ -2259,6 +2259,18 @@ FoldingRule BitReverseScalarOrVector() {
...
+        (constants[0]->AsVectorConstant() &&
+         std::any_of(constants[0]->AsVectorConstant()->GetComponents().begin(),
+                     constants[0]->AsVectorConstant()->GetComponents().end(),
+                     [](const analysis::Constant* c) {
+                       return c->AsNullConstant() != nullptr;
+                     }))) {
+      // We won't fold if the constant is null constant or any component is a
+      // null constant
```
Comment (s-perron) [inline source/opt/folding_rules.cpp]: Why not? What fails if you have a null?

File: test/opt/fold_test.cpp [hunk 4]
```diff
@@ -1413,6 +1437,11 @@ TEST_P(UIntVectorInstructionFoldingTest, Case) {
   Instruction* inst;
   std::tie(context, inst) =
       FoldInstruction(tc.test_body, tc.id_to_fold, SPV_ENV_UNIVERSAL_1_1);
+  // if inst is null, the instruction was not folded.
+  if (inst == nullptr) {
+    EXPECT_EQ(tc.expected_result.size(), 0);
+    return;
+  }
```
Comment (s-perron) [inline test/opt/fold_test.cpp]: I don't know if we want this to be forced like this. Can you explain why we want to do this?

Comment (luciechoi) [reply test/opt/fold_test.cpp]: (This was originally for checking if the instruction was folded or not, but removed as it's no longer applicable)

File: source/opt/folding_rules.cpp (hunk 2 above)
Comment (luciechoi) [reply source/opt/folding_rules.cpp]: Sounds good, updated the code to support folding `OpConstantNull` and added tests. PTAL!

File: source/opt/folding_rules.cpp [hunk 5]
```diff
@@ -182,6 +182,13 @@ std::vector<uint32_t> GetWordsFromNumericScalarOrVectorConstant(
       words.insert(words.end(), comp_in_words.begin(), comp_in_words.end());
     }
     return words;
+  } else if (c->AsNullConstant()) {
+    uint32_t num_elements = 1;
+    if (const auto* vec_type = c->type()->AsVector()) {
+      num_elements = vec_type->element_count();
+    }
+    // 16, 32 or 64 bit elements are all rounded up to 1 word.
+    return std::vector<uint32_t>(num_elements, 0);
```
Comment (s-perron) [inline source/opt/folding_rules.cpp:229]: Does this work with 64-bit types? an int64 might need 2 words per element, right?

**Comment by luciechoi** [comment]
Thank you for the thorough review, will merge now.
